packages/@sanity/schema/src/sanity/extractSchema.ts

```diff
@@ -0,0 +1,530 @@
+function gatherFields(type: SanitySchemaType | ObjectSchemaType): ObjectField[] {
+  if ('fields' in type) {
+    return type.type ? gatherFields(type.type).concat(type.fields) : type.fields
+  }
+
+  return []
+}
+
+// Traverse the type tree and check if the type or any of its subtypes are of the given type
+function isType(
+  typeDef: SanitySchemaType | ObjectField | ObjectFieldType,
+  typeName: string,
+): boolean {
+  let type: SchemaType | ObjectField | ObjectFieldType | undefined = typeDef
+  while (type) {
+    if (type.name === typeName || (type.type && type.type.name === typeName)) {
+      return true
+    }
+
+    type = type.type
+  }
+  return false
+}
+
+// Traverse the type tree and return the "last" type, ie deepest type in the tree
+function lastType(typeDef: SanitySchemaType): SanitySchemaType | undefined {
+  let type: SchemaType | ObjectField | ObjectFieldType | undefined = typeDef
+  while (type) {
+    if (!type.type) {
+      return type
+    }
+    type = type.type
+  }
+
+  return undefined
+}
+
+// Sorts the types by their dependencies by using a topological sort depth-first algorithm.
+function sortByDependencies(compiledSchema: SchemaDef): string[] {
+  const seen = new Set<SanitySchemaType>()
+
+  // Walks the dependencies of a schema type and adds them to the dependencies set
+  function walkDependencies(
+    schemaType: SanitySchemaType,
+    dependencies: Set<SanitySchemaType>,
+  ): void {
+    if (seen.has(schemaType)) {
+      return
+    }
+    seen.add(schemaType)
+
+    if ('fields' in schemaType) {
+      for (const field of gatherFields(schemaType)) {
+        let schemaTypeName: string | undefined
+        if (schemaType.type.type) {
+          schemaTypeName = field.type.type.name
+        } else if ('jsonType' in schemaType.type) {
+          schemaTypeName = field.type.jsonType
+        }
+
+        if (
+          schemaTypeName === 'document' ||
+          schemaTypeName === 'object' ||
+          schemaTypeName === 'block'
+        ) {
+          if (isReferenceType(field.type)) {
+            field.type.to.forEach((ref) => dependencies.add(ref.type))
+          } else {
+            dependencies.add(field.type)
+          }
+        }
+        walkDependencies(field.type, dependencies)
+      }
+    } else if ('of' in schemaType) {
+      for (const item of schemaType.of) {
+        walkDependencies(item, dependencies)
+      }
+    }
+  }
+  const dependencyMap = new Map<SanitySchemaType, Set<SanitySchemaType>>()
+  compiledSchema.getTypeNames().forEach((typeName) => {
+    const schemaType = compiledSchema.get(typeName)
+    if (schemaType === undefined || schemaType.type === null) {
+      return
+    }
+    const dependencies = new Set<SanitySchemaType>()
+
+    walkDependencies(schemaType, dependencies)
+    dependencyMap.set(schemaType, dependencies)
+  })
+
+  // Sorts the types by their dependencies
+  const typeNames: string[] = []
+  const tempMark = new Set<SanitySchemaType>()
```

Comment:
could we find a better name, e.g. `currentlyVisiting` and add a comment explaining that it's for detecting cyclic dependencies?